### ISSUE TITLE
[feature request] allow user to combine two or more tasks

### ISSUE THREAD

**Issue by arooni** [GothenburgBitFactory/taskwarrior#2198]
#### To request a feature...
task combine 22 23 165
Would collapse descriptions/notes from tasks from the second argument on to the first task listed.   Would be useful when I do a search across a few different tasks and realize they are really related to the first task.

**Comment by heiderich** [comment]
While I can partially understand this feature request, I do not see how it should be implemented in practice. While descriptions and annotations might simply be concatenated, it is not clear to me how to handle due and wait dates for instance.

**Comment by arooni** [comment]
over ride any conflicts with the first task's information.

**Comment by JonasDralle** [comment]
Personally I would also benefit from the ability of _splitting_ tasks. I often see that some sort of action on my to do list is a _multi-step project_. I then need to create more tasks out of this one task. Usually these resulting tasks have a lot in common (usually they share the same project tag, same deadline, same review date, sometimes the same annotations, ... etc)

Is there already an issue open on this?

**Comment by JonasDralle** [comment]
> over ride any conflicts with the first task's information.

In what ways are your tasks combined? Do the tasks share the same `project` or have `depend` relationships?

**Comment by arooni** [comment]
I think there is benefit to both splitting and combining tasks.  

What I find is I'll do a keyword search find a few tasks that really are the same task, and might have different notes/annotations per task which are really all the same task but instead of copying over the annotations and deleting the other tasks it would be nice to have a command that does it for me.

i think we could make this simpler if we just refuse to deal with any depends relationships in this command.  And use the project of the first task and overwrite everything else.   

Whether other people think this would be useful is a separate issue.  I'm also pretty open to how this would be implemented especially since I wouldn't be writing the code myself.  Beggars can't be choosers :)

**Comment by pbeckingham** [comment]
@JonasDralle there is a `duplicate` command which is equivalent to a split. Any further processing than duplication is difficult.

**Comment by pbeckingham** [comment]
@arooni the problems with joining tasks are:

- if I join three tasks, what does the description look like? Simple concatenation?
- `project` is a single attribute, so what are the rules for joining three tasks having two projects?
- If a UDA is an integer, what are the rules for joining?
- How are all the dates resolved?  Note that dates have internal consistencies to maintain (i.e. `entry` is always <= `end`

There are more issues, but these highlight the difficulty in automating joins.

As a general rule, this kind of command requires no core changes, and can be implemented as an external script with no loss in capability.  As such, it would be rejected for that reason alone.

**Comment by JonasDralle** [comment]
> @JonasDralle there is a `duplicate` command which is equivalent to a split. Any further processing than duplication is difficult.

whoa, nice

> As a general rule, this kind of command requires no core changes, and can be implemented as an external script with no loss in capability. As such, it would be rejected for that reason alone.

Yeah that's correct. Same would be true for any command tho

**Comment by JonasDralle** [comment]
> @JonasDralle there is a `duplicate` command which is equivalent to a split. Any further processing than duplication is difficult.

whoa, nice

**Comment by apkawel** [comment]
> @JonasDralle there is a `duplicate` command which is equivalent to a split. Any further processing than duplication is difficult.

I would like to see a `split` command too. When I think of a split, it is not so much a duplication of a task. Rather, it is a breakdown of a task into components. In my mind, `split` would add a new task (by prompting for entry of the description) and automatically (1) assign the same project name and (2) make the original task depend on the new task. (Perhaps the tags from the "parent task" would be automatically assigned as well.)

The use case would be GTD-like in nature — something like this: I have a task. When it shows up on the top of my task list, I realize it's not "actionable." In other words, there are more steps I need to take to accomplish it. (Hence the task will depend on those other steps.) If I split the task, the task will become +BLOCKED and I can start "drilling down" or "planning in reverse" (what David Allen calls the "natural planning model") by breaking the task down into concrete steps. Perhaps I iterate that process until I arrive at an actionable task (with the benefit of already having planned the next steps to take once I complete the broken-down task).

Example:

```
ID - Description - Project
1 - Have a great date at home with Sarah (Netflix and chill) - dating.sarah
---
task split 1 have a romantic dinner with my date {assigned as task 2}
task split 1 watch a movie on Netflix {assigned as task 3}
task split 2 cook the dinner {4}
task split 4 buy all the ingredients {5}
task split 5 decide what to cook {6}
task split 6 browse some cooking websites for recipe inspirations {7}
task split 3 prepare three suggested movies for us to watch so Sarah can pick one {8}
task split 8 browse Netflix for ideas
<etc.>
```

Doing it like this (particularly with automatic dependencies) would make project planning super fast and easy. In essence, the user would be laying out a dependency tree without the hassle of this:

<what does this depend on? oh yeah, and where is that task? let me look for it … what's the ID number? doh, what was the ID of the original task? now let me modify that original task to add a dependency with `task X modify depends:Y` … etc.> .

Pair the `split` command with a keybinding in taskwarrior-tui and you have an absolute powerhouse.

In sum, I came here looking for basically what @JonasDralle wanted originally (at least as I understand it). Perhaps this should be moved into a new issue / feature request?

Thank you for your work and consideration.

**Comment by tbabej** [comment]
I think this is a good suggestion, but perhaps we should implement this in the context of upcoming subtasks feature.

**Comment by apkawel** [comment]
> I think this is a good suggestion, but perhaps we should implement this in the context of upcoming subtasks feature.

I agree. In the meantime, I cobbled together a working system for splitting tasks by combining these:

-  @wbsch's `on-modify.blocks_attr.py` Python hook to create a `blocks` modification (https://gist.github.com/wbsch/a2f7264c6302918dfb30) (installation instructions in code comments);
- @kdheepak's `taskwarrior-tui` (with its shortcut feature) (https://github.com/kdheepak/taskwarrior-tui); and
-  a `split.sh` shell script that I threw together (below).

After installing `taskwarrior-tui` and `on-modify.blocks_attr.py`, add the following to `.taskrc`:

`uda.taskwarrior-tui.shortcuts.1=~/.config/taskwarrior-tui/shortcut-scripts/split.sh # Splits task by creating a duplicate task that blocks the task. This allows the user to break down the task into smaller chunks that are easier to accomplish. `

That code tells `taskwarrior-tui` to assign shortcut key `1` to execute `split.sh` using the selected task.

Then create `split.sh` with the following content and make it executable:

```
task rc.bulk=0 rc.confirmation=off rc.dependency.confirmation=off rc.recurrence.confirmation=off "$@" duplicate
task rc.bulk=0 rc.confirmation=off rc.dependency.confirmation=off rc.recurrence.confirmation=off "+LATEST" mod blocks:"$@"
read -p "Enter blocking task description: "  blockingtask
task "+LATEST" mod "$blockingtask"
```

That code:

-  duplicates the selected task; 
- modifies the duplicate task (+LATEST) to block the "parent" task (_i.e._, makes the parent task depend on the duplicate task); and 
- prompts the user to change the description of the duplicate task. 
- Then the user is returned to `taskwarrior-tui`. 

I'm sure there is a better or more-elegant way to do this, but I'm a novice and this was the best I could do. Also, you could probably modify it somehow to automatically return the selection marker in `taskwarrior-tui` to the top of the list, if that were the behavior you desired.

Now when you're in `taskwarrior-tui`, you can select a task and hit the shortcut key `1` to run the script. It works fairly well to start "breaking down" large tasks into more-manageable chunks by leaving a trail of "breadcrumbs" (in the form of smaller tasks and dependencies) so that you can follow your way back to your larger goals by completing the "split" or "broken down" tasks. This is similar to David Allen's "natural planning model" in GTD, and I think it works pretty well in practice.

 @JavaZauber , this might be of interest to you. I hope it helps, and maybe provides some inspiration for some kind of a "split" or "breakdown" feature in v3.0. (Or some improvement of the planned subtasks feature, as @tbabej mentioned.)

Cheers.
